File: services/frontend/src/stores/billing.ts
/**
 * Billing Zustand Store — Plans, subscriptions, invoices, payment methods.
 *
 * Manages billing state: current plan, available plans, invoices, payment methods,
 * feature access checks, and usage stats.
 */

import { create } from "zustand";
import { api } from "@/lib/api";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface Plan {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  price_monthly_cents: number;
  price_monthly_display: string;
  price_yearly_cents: number;
  price_yearly_display: string;
  max_agents: number | null;
  max_integrations: number | null;
  max_team_members: number | null;
  max_memory_nodes: number | null;
  max_graph_depth: number | null;
  max_reviews_per_day: number | null;
  daily_notes_history_days: number | null;
  progressive_summarization_layers: number;
  auto_linking_level: string;
  export_formats: string[];
  memory_analytics: string;
  features: Record<string, unknown>;
  sort_order: number;
  is_active: boolean;
}

export interface Subscription {
  id: string;
  status: string;
  billing_cycle: string;
  current_period_start: string | null;
  current_period_end: string | null;
  trial_end: string | null;
  canceled_at: string | null;
  stripe_subscription_id: string | null;
  paystack_subscription_code: string | null;
}

export interface CurrentPlanResponse {
  plan: Plan | null;
  subscription: Subscription | null;
  status: string;
  current_period_start: string | null;
  current_period_end: string | null;
}

export interface Invoice {
  id: string;
  amount_cents: number;
  amount_display: string;
  currency: string;
  status: string;
  period_start: string | null;
  period_end: string | null;
  paid_at: string | null;
  plan_name: string;
  billing_cycle: string;
  description: string;
}

export interface FeatureAccessResult {
  allowed: boolean;
  limit: number | null;
  current: number | null;
  plan_slug: string;
  plan_name: string;
  feature_key: string;
}

export interface UsageStats {
  agents: {
    limit: number | null;
    current: number | null;
    allowed: boolean;
  };
  integrations: {
    limit: number | null;
    current: number | null;
    allowed: boolean;
  };
  memory_nodes: {
    limit: number | null;
    current: number | null;
    allowed: boolean;
  };
  team_members: {
    limit: number | null;
    current: number | null;
    allowed: boolean;
  };
}

// ─── Store Interface ─────────────────────────────────────────────────────

interface BillingState {
  // Data
  currentPlan: CurrentPlanResponse | null;
  availablePlans: Plan[];
  invoices: Invoice[];
  usage: UsageStats | null;

  // Loading states
  isLoading: boolean;
  isUpgrading: boolean;
  error: string | null;

  // Success/feedback
  lastAction: string | null;

  // Actions
  fetchCurrentPlan: () => Promise<void>;
  fetchAvailablePlans: () => Promise<void>;
  fetchInvoices: () => Promise<void>;
  fetchUsage: () => Promise<void>;
  upgradePlan: (
    planSlug: string,
    billingCycle?: string,
    paymentMethodId?: string,
  ) => Promise<boolean>;
  downgradePlan: () => Promise<boolean>;
  cancelSubscription: () => Promise<boolean>;
  updatePaymentMethod: (paymentMethodId: string, provider?: string) => Promise<boolean>;
  checkFeatureAccess: (featureKey: string) => Promise<FeatureAccessResult>;
  clearError: () => void;
}

// ─── Store ─────────────────────────────────────────────────────────────

export const useBillingStore = create<BillingState>((set, get) => ({
  // ── State ──
  currentPlan: null,
  availablePlans: [],
  invoices: [],
  usage: null,
  isLoading: false,
  isUpgrading: false,
  error: null,
  lastAction: null,

  // ── Actions ──

  fetchCurrentPlan: async () => {
    set({ isLoading: true, error: null });

    try {
      const data = await api.get<CurrentPlanResponse>("/api/v1/billing/plan");
      set({ currentPlan: data, isLoading: false, lastAction: "fetch_plan" });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load plan";
      set({ isLoading: false, error: message });
    }
  },

  fetchAvailablePlans: async () => {
    try {
      const data = await api.get<Plan[]>("/api/v1/billing/plans");
      set({ availablePlans: data });
    } catch {
      // Non-critical
    }
  },

  fetchInvoices: async () => {
    try {
      const data = await api.get<Invoice[]>("/api/v1/billing/invoices");
      set({ invoices: data });
    } catch {
      // Non-critical
    }
  },

  fetchUsage: async () => {
    try {
      const data = await api.get<UsageStats>("/api/v1/billing/usage");
      set({ usage: data });
    } catch {
      // Non-critical
    }
  },

  upgradePlan: async (planSlug: string, billingCycle = "monthly", paymentMethodId?: string) => {
    set({ isUpgrading: true, error: null });

    try {
      const body: Record<string, unknown> = {
        plan_slug: planSlug,
        billing_cycle: billingCycle,
      };
      if (paymentMethodId) {
        body.payment_method_id = paymentMethodId;
      }

      const data = await api.post<CurrentPlanResponse>("/api/v1/billing/upgrade", body);
      set({
        currentPlan: data,
        isUpgrading: false,
        lastAction: `upgrade_to_${planSlug}`,
      });

      // Refresh available plans and usage
      get().fetchAvailablePlans();
      get().fetchUsage();

      return true;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to upgrade plan";
      set({ isUpgrading: false, error: message });
      return false;
    }
  },

  downgradePlan: async () => {
    set({ isUpgrading: true, error: null });

    try {
      const data = await api.post<CurrentPlanResponse>("/api/v1/billing/downgrade");
      set({
        currentPlan: data,
        isUpgrading: false,
        lastAction: "downgrade_to_free",
      });
      return true;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to downgrade plan";
      set({ isUpgrading: false, error: message });
      return false;
    }
  },

  cancelSubscription: async () => {
    set({ isUpgrading: true, error: null });

    try {
      const data = await api.post<CurrentPlanResponse>("/api/v1/billing/cancel");
      set({
        currentPlan: data,
        isUpgrading: false,
        lastAction: "cancel_subscription",
      });
      return true;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to cancel subscription";
      set({ isUpgrading: false, error: message });
      return false;
    }
  },

  updatePaymentMethod: async (paymentMethodId: string, provider = "stripe") => {
    set({ isUpgrading: true, error: null });

    try {
      await api.post("/api/v1/billing/payment-method", {
        payment_method_id: paymentMethodId,
        provider,
      });
      set({ isUpgrading: false, lastAction: "update_payment_method" });
      return true;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to update payment method";
      set({ isUpgrading: false, error: message });
      return false;
    }
  },

  checkFeatureAccess: async (featureKey: string) => {
    try {
      const data = await api.post<FeatureAccessResult>("/api/v1/billing/check-feature", {
        feature_key: featureKey,
      });
      return data;
    } catch {
      return {
        allowed: false,
        limit: null,
        current: null,
        plan_slug: "free",
        plan_name: "Free",
        feature_key: featureKey,
      };
    }
  },

  clearError: () => set({ error: null }),
}));

File: services/backend/app/api/v1/billing.py
"""
Billing API Endpoints — Plans, subscriptions, invoices, payment methods, webhooks.

From spec section 8.2:
- GET    /api/v1/billing/plan — Current plan
- POST   /api/v1/billing/upgrade — Change plan
- POST   /api/v1/billing/downgrade — Cancel/downgrade
- GET    /api/v1/billing/invoices — Invoice history
- POST   /api/v1/billing/payment-method — Update payment method
- POST   /api/v1/billing/webhook/stripe — Stripe webhook (public)
- POST   /api/v1/billing/webhook/paystack — Paystack webhook (public)
- GET    /api/v1/billing/plans — List available plans
- POST   /api/v1/billing/check-feature — Check plan feature access
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Header, Query, Request
from structlog import get_logger

from app.core.dependencies import CurrentUser, get_current_user, rate_limit
from app.core.exceptions import ValidationError
from app.services.billing import BillingService

logger = get_logger(__name__)

router = APIRouter()
billing_service = BillingService()


# ─── Plans ─────────────────────────────────────────────────────────────────────


@router.get("/plans")
async def list_plans(
    current_user: CurrentUser | None = None,
) -> list[dict[str, Any]]:
    """List all available plans with features and pricing."""
    return await billing_service.list_plans()


@router.get("/plan")
async def get_current_plan(
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Get the current user's plan and subscription status."""
    return await billing_service.get_plan(user_id=current_user.id)


# ─── Upgrade / Downgrade ──────────────────────────────────────────────────────


@router.post("/upgrade")
async def upgrade_plan(
    body: dict[str, Any],
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Upgrade or change the current user's plan.

    Body:
        plan_slug (str, required): Target plan slug ('pro', 'business')
        billing_cycle (str, default='monthly'): 'monthly' or 'yearly'
        payment_method (str, optional): 'stripe' or 'paystack'
        payment_method_id (str, optional): Payment source ID from provider
    """
    plan_slug = body.get("plan_slug")
    if not plan_slug:
        raise ValidationError(message="plan_slug is required")

    billing_cycle = body.get("billing_cycle", "monthly")
    payment_method = body.get("payment_method")
    payment_method_id = body.get("payment_method_id")

    return await billing_service.upgrade_plan(
        user_id=current_user.id,
        plan_slug=plan_slug,
        billing_cycle=billing_cycle,
        payment_method=payment_method,
        payment_method_id=payment_method_id,
    )


@router.post("/downgrade")
async def downgrade_plan(
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Schedule downgrade to free plan at the end of the current period."""
    return await billing_service.downgrade_plan(user_id=current_user.id)


@router.post("/cancel")
async def cancel_subscription(
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Cancel subscription at the end of the current period."""
    return await billing_service.cancel_subscription(user_id=current_user.id)


# ─── Invoices ──────────────────────────────────────────────────────────────────


@router.get("/invoices")
async def get_invoices(
    limit: int = Query(20, ge=1, le=100, description="Maximum invoices to return"),
    current_user: CurrentUser = Depends(get_current_user),
) -> list[dict[str, Any]]:
    """Get invoice history for the current user."""
    return await billing_service.get_invoices(
        user_id=current_user.id,
        limit=limit,
    )


# ─── Payment Method ────────────────────────────────────────────────────────────


@router.post("/payment-method")
async def update_payment_method(
    body: dict[str, Any],
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Update the payment method on file.

    Body:
        payment_method_id (str, required): Payment method ID from provider
        provider (str, default='stripe'): 'stripe' or 'paystack'
    """
    payment_method_id = body.get("payment_method_id")
    if not payment_method_id:
        raise ValidationError(message="payment_method_id is required")

    provider = body.get("provider", "stripe")

    return await billing_service.update_payment_method(
        user_id=current_user.id,
        payment_method_id=payment_method_id,
        provider=provider,
    )


# ─── Feature Access Check ─────────────────────────────────────────────────────


@router.post("/check-feature")
async def check_feature_access(
    body: dict[str, Any],
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Check if the current user's plan allows access to a feature.

    Body:
        feature_key (str, required): Feature key to check
            (e.g., 'max_agents', 'max_integrations', 'private_marketplace')

    Used by other services (agents, integrations, brain) to enforce plan limits.
    """
    feature_key = body.get("feature_key")
    if not feature_key:
        raise ValidationError(message="feature_key is required")

    return await billing_service.check_plan_feature_access(
        user_id=current_user.id,
        feature_key=feature_key,
    )


# ─── Usage Stats ──────────────────────────────────────────────────────────────


@router.get("/usage")
async def get_usage_stats(
    current_user: CurrentUser = Depends(get_current_user),
) -> dict[str, Any]:
    """Get the current user's usage statistics against plan limits.

    Returns counts for agents, integrations, memory nodes, and their limits.
    """
    usage: dict[str, Any] = {
        "agents": {},
        "integrations": {},
        "memory_nodes": {},
        "team_members": {},
    }

    # Fetch each feature check
    for feature_key in ("max_agents", "max_integrations", "max_memory_nodes", "max_team_members"):
        key_name = feature_key.replace("max_", "")
        result = await billing_service.check_plan_feature_access(
            user_id=current_user.id,
            feature_key=feature_key,
        )
        usage[key_name] = {
            "limit": result["limit"],
            "current": result["current"],
            "allowed": result["allowed"],
        }

    return usage


# ─── Webhooks (Public, no auth) ────────────────────────────────────────────────


@router.post("/webhook/stripe")
async def stripe_webhook(
    request: Request,
    stripe_signature: str | None = Header(default=None, alias="Stripe-Signature"),
) -> dict[str, Any]:
    """Handle incoming Stripe webhook events.

    This endpoint is public (no auth) — security is provided by
    Stripe's webhook signature verification.
    """
    payload = await request.body()
    sig_header = stripe_signature or ""

    return await billing_service.handle_stripe_webhook(
        payload=payload,
        sig_header=sig_header,
    )


@router.post("/webhook/paystack")
async def paystack_webhook(
    request: Request,
    x_paystack_signature: str | None = Header(default=None, alias="x-paystack-signature"),
) -> dict[str, Any]:
    """Handle incoming Paystack webhook events.

    This endpoint is public (no auth) — security is provided by
    Paystack's HMAC-SHA512 webhook signature verification.
    """
    payload = await request.body()

    return await billing_service.handle_paystack_webhook(
        payload=payload,
    )
